pkg/csr/csr.go

```diff
@@ -0,0 +1,219 @@
+// Config holds the information required to authorize and approve a CSR
+type Config struct {
+	k8sClientSet      *kubernetes.Clientset
+	instanceAddresses []string
+}
+
+// NewConfig returns a pointer to the Config
+func NewConfig(k8sclientSet *kubernetes.Clientset, instanceAddresses []string) (*Config, error) {
+	if len(instanceAddresses) == 0 || k8sclientSet == nil {
+		return nil, errors.New("instance address list or kubernetes clientset should not be nil")
+	}
+	return &Config{k8sclientSet, instanceAddresses}, nil
+}
+
+// Authorize returns true if the CSR request contents can be validated
+func (c *Config) Authorize(csr *certificatesv1.CertificateSigningRequest) (bool, error) {
+	parsedCSR, err := parseCSR(csr)
+	if err != nil {
+		return false, errors.Wrapf(err, "error parsing CSR %s", csr.Name)
+	}
+
+	dnsAddr := strings.TrimPrefix(parsedCSR.Subject.CommonName, nodePrefix)
+	if dnsAddr == "" {
+		return false, errors.Errorf("CSR subject name does not contain DNS address %s", csr.Name)
+	}
+
+	// check DNS of node requesting CSR matches instance IP/DNS address
+	var hasMatch bool
+	for _, instanceAdd := range c.instanceAddresses {
+		if areEqualAddresses(instanceAdd, dnsAddr) {
+			hasMatch = true
+			break
+		}
+	}
+	if !hasMatch {
+		return false, errors.Errorf("CSR is not from an authorized source %s", csr.Name)
+	}
+
+	if isNodeClientCert(csr, parsedCSR) {
+		// this is done before the instance becomes a node
+		return c.authorizeNodeClientCSR(dnsAddr)
+	}
+
+	return c.authorizeKubeletServingCSR(csr, parsedCSR)
+}
+
+// authorizeKubeletServingCSR authorizes a kubelet serving CSR
+func (c *Config) authorizeKubeletServingCSR(req *certificatesv1.CertificateSigningRequest, csr *x509.CertificateRequest) (bool, error) {
+	if req == nil || csr == nil {
+		return false, errors.New("CSR should not be nil")
+	}
+
+	// Check groups, we need at least: system:nodes, system:authenticated
+	if len(req.Spec.Groups) < 2 {
+		return false, errors.Errorf("CSR contains too few groups: %s", req.Name)
+	}
+	groupSet := sets.NewString(req.Spec.Groups...)
+	if !groupSet.HasAll(nodeGroup, systemPrefix) {
+		return false, errors.Errorf("CSR does not contain required groups: %s", req.Name)
+	}
+
+	// Check usages include: digital signature, key encipherment and server auth
+	if !hasExactUsages(req, kubeletServerUsages) {
+		return false, errors.Errorf("CSR does not contain required usages: %s", req.Name)
+	}
+
+	var hasOrg bool
+	for i := range csr.Subject.Organization {
+		if csr.Subject.Organization[i] == nodeGroup {
+			hasOrg = true
+			break
+		}
+	}
+	if !hasOrg {
+		return false, errors.Errorf("CSR does not contain required organization: %s", req.Name)
+	}
+	return true, nil // Approve node server cert
+}
+
+// authorizeNodeClientCSR validates a kubelet client CSR by verifying if it already
+// exists as a node because client CSR's are issued by an instance before becoming a node.
+func (c *Config) authorizeNodeClientCSR(nodeName string) (bool, error) {
```

Comment:
I don't see this function doing any authorization. It seems to be checking if a node object exists.